keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,420 @@
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+High Performance Computing / Machine Learning users usually want some control over how jobs are being created.  A Startup policy would allow a user to specify which status a replicated job should have before starting other replicated jobs in a jobSet.
+
+### Goals
+
+- Add a startup policy for JobSet
+
+### Non-Goals
+
+- Complicated workflow specifications
+  - This means that we will support simple dependecies rather than processing a full DAG.
+  - Projects like argo-workflows are a better use case for this.
+- Starting Jobs based on Ready will be at the ReplicatedJob level not at the Job level.
+  - ReplicatedJobs will be classified as ready if all jobs are ready.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I have one ReplicatedJob that functions as the driver and a series of worker ReplicatedJobs.  
+I want the driver to be ready before I start the worker replicated jobs.  
+This is useful for HPC/AI/ML workloads as some frameworks have a driver pod that must be ready before the workers are started.  
+
+The API below would fit my purpose.
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: driver-ready-worker-start
+spec:
+  startupPolicy:
+     replicatedJobsStartInOrder: true
+  replicatedJobs:
+  - name: driver
+    replicas: 1
+    template:
+      spec:
+        # Set backoff limit to 0 so job will immediately fail if any pod fails.
+        backoffLimit: 0 
+        completions: 1
+        parallelism: 1
+        template:
+          spec:
+            containers:
+            - name: driver
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10000
+  - name: workers
+    replicas: 1
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+```
+
+#### Story 2
+
+As a user, I want my replicated jobs to start in order.
+
+One area that this could be useful is if someone has a redis replicatedJob, followed by a driver replicatedJob, followed by a worker replicated job.
```

Comment:
Do people use Job to deploy redis?

Reply:
@vsoch has pointed out to me workflows like https://ieeexplore.ieee.org/abstract/document/9561187?casa_token=aoAoa11fcfoAAAAA:jGlCi6-v5KObaSSlOJooCFCtlpSAQJJkEV8zrVOCTDJfBcf2WihKaDLNTgnd_PqwWQPdQA0 that may require a database/service be deployed as part of a batch job. Maybe a MessageQueue is more likely but there are some examples in the wild.

You are correct that most likely someone should use a Deployment for Redis but I have heard of people wanting batch jobs that utilize some kind of database in the lifecycle of their job and have it complete once the other tasks are complete.

Reply:
That is correct. What distinguishes these services is that they are only needed for the scoped run of the job - it's a strategy for worker communication that can be cleaned up with them.

Reply:
I updated the user story to say its a message queue rather than redis.